nemo/collections/nlp/models/multi_label_intent_slot_classification/multi_label_intent_slot_classification_model.py

```diff
@@ -0,0 +1,406 @@
+
+            for batch in infer_datalayer:
+                input_ids, input_type_ids, input_mask, loss_mask, subtokens_mask = batch
+
+                intent_logits, slot_logits = self.forward(
+                    input_ids=input_ids.to(device),
+                    token_type_ids=input_type_ids.to(device),
+                    attention_mask=input_mask.to(device),
+                )
+
+                # predict intents and slots for these examples
+                # intents
+                probabilities.append(torch.sigmoid(intent_logits))
+
+            probabilities = torch.cat(probabilities)
+
+        finally:
+            # set mode back to its original value
+            self.train(mode=mode)
+
+        return probabilities
+
+    def optimize_threshold(self, test_ds):
+        input_file = f"{self.data_dir}/dev.tsv"
+
+        with open(input_file, "r") as f:
+            input_lines = f.readlines()[1:]  # Skipping headers at index 0
+
+        dataset = list(input_lines)
+
+        metrics_labels, sentences = [], []
+
+        for input_line in dataset:
+            sentence = input_line.strip().split("\t")[0]
+            sentences.append(sentence)
+            parts = input_line.strip().split("\t")[1:][0]
+            parts = list(map(int, parts.split(",")))
+            parts = [1 if label in parts else 0 for label in range(len(self.cfg.data_desc.intent_labels))]
+            metrics_labels.append(parts)
+
+        # Retrieve class probabilities for each sentence
+        intent_probabilities = self.prediction_probabilities(sentences, test_ds)
+
+        metrics_dict = {}
+        # Find optimal logits rounding threshold for intents
+        for i in np.arange(0.5, 0.96, 0.01):
+            predictions = tensor2list((intent_probabilities >= i).int())
+            precision = precision_score(metrics_labels, predictions, average='micro')
+            recall = recall_score(metrics_labels, predictions, average='micro')
+            f1 = f1_score(metrics_labels, predictions, average='micro')
+            metrics_dict[i] = [precision, recall, f1]
+
+        max_precision = max(metrics_dict, key=lambda x: metrics_dict[x][0])
+        max_recall = max(metrics_dict, key=lambda x: metrics_dict[x][1])
+        max_f1_score = max(metrics_dict, key=lambda x: metrics_dict[x][2])
+
+        logging.info(
+            f'Maximum Threshold for F1-Score: {max_f1_score}, [Precision, Recall, F1-Score]: {metrics_dict[max_f1_score]}'
+        )
+        logging.info(
+            f'Maximum Threshold for Precision: {max_precision}, [Precision, Recall, F1-Score]: {metrics_dict[max_precision]}'
+        )
+        logging.info(
+            f'Maximum Threshold for Recall: {max_recall}, [Precision, Recall, F1-Score]: {metrics_dict[max_recall]}'
+        )
+
+        if metrics_dict[max_f1_score][2] > self.max_f1:
+            self.max_f1 = metrics_dict[max_f1_score][2]
+            self.threshold = max_f1_score
```

Comment:
Since you set a threshold here, I think you should also add a line of logging to tell the user that. Something like "Setting threshold to <whatever>"